### PR TITLE
Bug 1943315: daemon: make node drain logic smarter for ICSP changes

### PR DESCRIPTION
**Pull request by sinnykumari**
Today, we perform drain for any changes made into container registry config file /etc/containers/registries.conf.
    
- To avoid users workload disruption, we can safely skip drain for some of the changes made into the config file. Such as:
    1. A new mirror is added to an existing registry that has `mirror-by-digest-only=true`
    2. A new registry has been added that has `mirror-by-digest-only=true`
    
- If any other unsafe changes are detected in registry conf, MCD will default to perform node drain.

- Also added unit tests.

### CODE REVIEW

**Comment by sinnykumari** [comment]
/cc @mrunalp @umohnani8 
can you please also take a look to make sure we are not adding any false positive or negative.

**Comment by kikisdeliveryservice** [comment]
Wow Sinny. :heart: 

Is this really a bug though?  Feels like it should've gone through other channels...

**Comment by sinnykumari** [comment]
> Wow Sinny. heart
> 
> Is this really a bug though? Feels like it should've gone through other channels...

Yeah, it is not a bug but was good to have in 4.8 if possible for  :)

File: pkg/daemon/drain_test.go [hunk 1]
```diff
@@ -0,0 +1,343 @@
...
+		},
+		{
+			// skip drain: only additional unqualified-search-registries has been added
+			actions:        []string{postConfigChangeActionReloadCrio},
+			oldConfig:      machineConfigs["mc3"],
+			newConfig:      machineConfigs["mc4"],
+			expectedAction: false,
+		},
+		{
+			// perform drain: prefix value of one or more registry has changed
```
Comment (yuqi-zhang) [inline pkg/daemon/drain_test.go:312]: Very minor, comparing 7 to 1, I don't think `prefix` value was changed, instead a registries.mirror was added with mirror-by-digest-only true (this is fine) but the other registry's mirror-by-digest-only changed from true to false, which is an unsafe change covered by the last test in the list. Did you mean to use other MCs here?

**Review by yuqi-zhang — APPROVED** [review]
Overall looks great, logic was very easy to follow, and amazing test coverage! Just one minor question below, otherwise will LGTM after giving others a chance to take a look. Thanks for working on this!

File: pkg/daemon/drain_test.go (hunk 1 above)
Comment (sinnykumari) [reply pkg/daemon/drain_test.go:328]: good catch Jerry! fixed it by adding a new MC that has only prefix change.

**Comment by umohnani8** [comment]
LGTM. Thanks for working on this!

**Pull request merged**